File: framework/languages/dotnet/samples/TicTacToe/Client/TicTacToeClientArguments.cs

namespace TicTacToe.Client;

internal static class TicTacToeClientArguments
{
    public static TicTacToeClientOptions Parse(string[] args)
    {
        var defaults = TicTacToeClientOptions.CreateDefault();
        var apiUrl = ReadOption(args, "--api-url") ?? defaults.ApiUrl.ToString();
        var gameName = ReadOption(args, "--game-name") ?? defaults.GameName;
        var xActorId = ReadOption(args, "--x-actor-id") ?? defaults.XActorId;
        var oActorId = ReadOption(args, "--o-actor-id") ?? defaults.OActorId;

        return defaults with
        {
            ApiUrl = new Uri(apiUrl),
            GameName = gameName,
            XActorId = xActorId,
            OActorId = oActorId,
        };
    }

    private static string? ReadOption(string[] args, string name)
    {
        var index = Array.IndexOf(args, name);
        if (index < 0)
        {
            return null;
        }

        if (index + 1 >= args.Length)
        {
            throw new ArgumentException($"Missing value for '{name}'.");
        }

        return args[index + 1];
    }
}

File: framework/languages/java/samples/java/TicTacToe/Client/src/main/java/systems/zlink/samples/tictactoe/client/TicTacToeClientArguments.java
package systems.zlink.samples.tictactoe.client;

public final class TicTacToeClientArguments {
    private TicTacToeClientArguments() {
    }

    public static TicTacToeClientOptions parse(String[] args) {
        TicTacToeClientOptions defaults = TicTacToeClientOptions.createDefault();
        return new TicTacToeClientOptions(
            readOption(args, "--game-name", defaults.gameName()),
            readOption(args, "--host-access-token", defaults.hostAccessToken()),
            readOption(args, "--guest-access-token", defaults.guestAccessToken()),
            readOption(args, "--api-endpoint", defaults.apiEndpoint()),
            readOption(args, "--play-endpoint", defaults.playEndpoint()));
    }

    private static String readOption(
        String[] args,
        String name,
        String defaultValue) {
        for (int index = 0; index < args.length; index++) {
            if (!args[index].equals(name)) {
                continue;
            }
            if (index + 1 >= args.length) {
                throw new IllegalArgumentException("Missing value for '" + name + "'.");
            }
            return args[index + 1];
        }
        return defaultValue;
    }
}
